### PR TITLE
SubBridge dynamic trader

### PR DESCRIPTION
**Pull request by tolak**
The origin [PR](https://github.com/Phala-Network/khala-parachain/pull/92) to implement dynamic trader is not compatible with the new SubBridge code, so I picked those stuff to a new PR.

With the dynamic trader introduced, an execution price would be given when registering a new asset. The execution price indicates how much that asset needs to pay per second on our network when buying the resource, which has the same meaning as the returned value of [pha_per_second()](https://github.com/Phala-Network/khala-parachain/blob/12d76b52cf90084a742618aee36b95893327b06a/runtime/khala/src/constants.rs#L73).

This PR would contain storage migration on pallet `chainbridge` and pallet `assets-registry`. They are:
- Remove the fee scale from `chainbridge::BridgeFee`, every ChainBridge only has a constant min fee set.
- Add `execution_price` into asset registry info, which would be used to calculate fee when sending a crosschain transfer of that asset

### CODE REVIEW

File: pallets/subbridge/src/dynamic_trader.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+where
+	WeightPerSecond: Get<u64>,
+	FungibleAssetsInfo: GetAssetRegistryInfo<FungibleAssetId>,
+	R: TakeRevenue,
+{
+	fn new() -> Self {
+		Self(0, 0, None, PhantomData)
+	}
+
+	fn buy_weight(&mut self, weight: Weight, payment: Assets) -> Result<Assets, XcmError> {
```
Comment (h4x3rotab) [inline pallets/subbridge/src/dynamic_trader.rs:33]: Please double check if saturating operations are good. Maybe better to use checked operations?

File: pallets/assets-registry/src/lib.rs [hunk 2]
```diff
@@ -473,7 +525,35 @@ pub mod pallet {
...
+
+		fn asset_ed(decimals: u8) -> <T as pallet_assets::Config>::Balance {
+			let native_ed: u128 = T::MinBalance::get().into();
+			if decimals >= 12 {
+				native_ed
+					.saturating_mul(10u128.saturating_pow(decimals as u32 - 12))
+					.into()
+			} else {
+				// + 1 make sure min balance always > 0
+				(native_ed.saturating_div(10u128.saturating_pow(12 - decimals as u32)) + 1).into()
```
[the remark this answers is not in the document]
Comment (tolak) [reply pallets/assets-registry/src/lib.rs:548]: Done

File: pallets/subbridge/src/migration.rs [hunk 3]
```diff
@@ -1,13 +1,10 @@
 use crate::{chainbridge, xcmbridge, xtransfer};
 
-pub mod subbridge_migration {
+mod subbridge_v3_migration_common {
```
[the remark this answers is not in the document]
Comment (tolak) [reply pallets/subbridge/src/migration.rs:3]: Done

File: pallets/subbridge/src/xcmbridge.rs [hunk 4]
```diff
@@ -393,7 +367,8 @@ pub mod pallet {
 			let (dest_location, beneficiary) =
 				Pallet::<T>::extract_dest(&dest).ok_or(Error::<T>::IllegalDestination)?;
 
-			let fee = Pallet::<T>::get_fee(&asset).ok_or(Error::<T>::AssetNotFound)?;
+			// We can not grantee asset has been supported by destination
```
[the remark this answers is not in the document]
Comment (tolak) [reply pallets/subbridge/src/xcmbridge.rs:370]: Done

File: pallets/assets-registry/src/lib.rs [hunk 5]
```diff
@@ -473,7 +525,35 @@ pub mod pallet {
...
+	{
+		fn default_price(native_price: u128, decimals: u8) -> u128 {
+			if decimals >= 12 {
+				native_price.saturating_mul(10u128.saturating_pow(decimals as u32 - 12))
+			} else {
+				native_price.saturating_div(10u128.saturating_pow(12 - decimals as u32))
+			}
+		}
+
+		fn asset_ed(decimals: u8) -> <T as pallet_assets::Config>::Balance {
```
[the remark this answers is not in the document]
Comment (tolak) [reply pallets/assets-registry/src/lib.rs:540]: Done. ED can be configured by authorized extrisic [force_asset_status](https://github.com/paritytech/substrate/blob/fe5bf49290d166b9552f65e751d46ec592173ebd/frame/assets/src/lib.rs#L1125) provided by pallet-asset

File: pallets/subbridge/src/chainbridge.rs [hunk 6]
```diff
@@ -2321,8 +2296,15 @@ pub mod pallet {
 
 			ParaA::execute_with(|| {
 				let dest_chain: u8 = 2;
+				let bridge_fee = 2;
 				let test_asset_location = MultiLocation::new(1, X1(GeneralKey(b"test".to_vec())));
-				assert_ok!(ChainBridge::update_fee(Origin::root(), 2, 0, dest_chain));
+				let unregistered_asset_location =
```
[the remark this answers is not in the document]
Comment (tolak) [reply pallets/subbridge/src/chainbridge.rs:2301]: Done

File: pallets/assets-registry/src/lib.rs [hunk 7]
```diff
@@ -678,5 +778,169 @@ pub mod pallet {
 				assert_eq!(AssetsRegistry::id(&para_b_location), None);
 			});
 		}
+
+		#[test]
+		fn test_non_registered_asset_price() {
+			new_test_ext().execute_with(|| {
+				let para_a_location: MultiLocation = MultiLocation {
```
[the remark this answers is not in the document]
Comment (tolak) [reply pallets/assets-registry/src/lib.rs:785]: Done

File: pallets/subbridge/src/chainbridge.rs [hunk 8]
```diff
@@ -815,29 +797,16 @@ pub mod pallet {
...
-				let decimals = if T::NativeAssetChecker::is_native_asset(asset) {
-					12
-				} else {
-					let id = T::AssetsRegistry::id(&location.clone())?;
-					T::AssetsRegistry::decimals(&id).unwrap_or(12)
-				};
-				if let Some(fee_in_pha) =
-					Self::estimate_fee_in_pha(chain_id, (Self::to_e12(amount, decimals)).into())
-				{
+			if let Some((location, _amount)) = Self::extract_fungible(asset.clone()) {
```
[the remark this answers is not in the document]
Comment (tolak) [reply pallets/subbridge/src/chainbridge.rs:800]: Awesome! Done.

File: pallets/subbridge/src/dynamic_trader.rs (hunk 1 above)
Comment (tolak) [reply pallets/subbridge/src/dynamic_trader.rs:33]: The above line ```let unused = payment.checked_sub((id, amount).into()).map_err(|_| XcmError::TooExpensive)?;``` already has overflow check, following `saturating_xxx` can work properly

Comment (h4x3rotab) [reply pallets/subbridge/src/dynamic_trader.rs:33]: In such case, I think you can just use plain operator (`+`, `-`)

Comment (tolak) [reply pallets/subbridge/src/dynamic_trader.rs:33]: Just follow the upstream [impl](https://github.com/paritytech/polkadot/blob/2c285771342a84e13dab696ef86902f73d4fbd4c/xcm/xcm-builder/src/weight.rs#L158)

**Review by h4x3rotab — APPROVED** [review]
LGTM

**Pull request merged**